docs/proposals/TLSPolicy.md

```diff
@@ -0,0 +1,83 @@
+
+## Use Cases
+
+**Choosing a TLS Provider**: As an admin, I want to select a specific ACME provider for the TLS certificates in a particular gateway.
+
+**Setting the TLS Certificate Renewal Window**: As an admin, I need to define the renewal period for TLS certificates (e.g., every 90 days) and a lead time for renewals (e.g., 15 days before expiration).
+
+**Choosing a TLS Strategy**: As an admin, I want to employ different strategies for managing certificates (e.g., multiple host certificate or per host certificate).
+
+**Choosing a TLS Strategy**: As an admin, I want to be able to apply enterprise security standards to certificates generated by the MGC. I want to be able to specify certificate key size, algorithm and encoding.
+
+
+## Current implementation
+
+
+
+## Goals
+
+- Design an API that allows admins to handle Automatic TLS for any given gateway.
+- Facilitate easy selection of a TLS provider.
+- Allow admins to set the renewal window for TLS certificates.
+- Enable the use of different TLS strategies.
+- Consider additional TLS configurations such as enforcing HTTPS listeners and supporting wildcard hosts.
+
+## Non-Goals
+
+- Detailed implementation of the API.
+- Developing a working prototype of the API.
+- Identifying specific technologies or tools for implementing the API.
+
+## Terms
+
+- **Automatic TLS**: Automatic management of TLS certificates for gateways.
+- **TLS Provider**: The entity providing TLS certificates.
+- **TLS Strategy**: The method used for managing and distributing TLS certificates across hosts.
+
+## Proposal
+
+The proposed solution is the TLSPolicy API. It allows administrators to select a TLS provider, define a renewal window for TLS certificates, and choose a TLS strategy. The API will support multiple host certificates and per host certificates.
+
+The API will also consider additional TLS configurations, such as enforcing only HTTPS listeners and supporting wildcard hosts. This means any gateways using this class would have any non-HTTPS listeners flagged, removed, or not synced, etc. 
+
+**Note** This proposal effectively moves us away from having a "default" TLS policy, baked into the MGC. From here on, unless a TLSPolicy is defined, MGC will not do anything with TLS, and control will remain with the user to solve.
+
+## Gateway Interaction
+
+The TLSPolicy API will interact with Gateway operations, such as adding and removing TLS configurations for a listener based on the host. This allows for dynamic management of TLS configurations for each Gateway and its listeners. The API will utilize the existing Gateway methods for adding, removing, and checking TLS status.
+
+## Example Policies
+
+1. **Policy**
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: TLSPolicy
+spec:
+  targetRef: # defaults to gateway gvk and currrent namespace
+    name: gateway-name
+  duration: 2160h # 90d - certificate validity period
```

Comment:
I'm assuming these fields map to fields in the [Certificate spec](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.Certificate)?

Would be good to have sensible defaults (e.g. what's used now) to ease the 'getting started' experience.

Open question if the TLSPolicy should embed the types directly from cert-manager (hard dependency) or copy them.

Reply:
agreed @david-martin. Yeah, I've made these more closely map to the certificate spec recently. I'm hoping a working PoC will help me flush out an answer on embedding types from cert-manager or copy them, I'm unsure right now.